src/main/java/picard/illumina/ClusterDataToSamConverter.java

```diff
@@ -207,49 +242,90 @@ private SAMRecord createSamRecord(final ReadData readData, final String readName
-                molecularIndexes.add(new String(cluster.getRead(molecularBarcodeIndices[i]).getBases()).replace('.', 'N'));
-                molecularIndexQualities.add(SAMUtils.phredToFastq(cluster.getRead(molecularBarcodeIndices[i]).getQualities()));
+            for (int molecularBarcodeIndice : molecularBarcodeIndices) {
+                molecularIndexes.add(convertMissingToNoCall(new String(cluster.getRead(molecularBarcodeIndice).getBases())));
+                molecularIndexQualities.add(SAMUtils.phredToFastq(cluster.getRead(molecularBarcodeIndice).getQualities()));
             }
-        } else {
-            molecularIndexes        = Collections.emptyList();
-            molecularIndexQualities = Collections.emptyList();
         }
 
         final SAMRecord firstOfPair = createSamRecord(
-            cluster.getRead(templateIndices[0]), readName, cluster.isPf(), true, unmatchedBarcode, molecularIndexes, molecularIndexQualities);
+                cluster.getRead(templateIndices[0]),
+                readName,
+                cluster.isPf(),
+                true,
+                unmatchedBarcode,
+                barcodeQuality,
+                molecularIndexes,
+                molecularIndexQualities);
         ret.records[0] = firstOfPair;
 
+
         SAMRecord secondOfPair = null;
 
-        if(isPairedEnd) {
-            secondOfPair  = createSamRecord(
-                cluster.getRead(templateIndices[1]), readName, cluster.isPf(), false, unmatchedBarcode, molecularIndexes, molecularIndexQualities);
+        if (isPairedEnd) {
+            secondOfPair = createSamRecord(
+                    cluster.getRead(templateIndices[1]),
+                    readName,
+                    cluster.isPf(),
+                    false,
+                    unmatchedBarcode,
+                    barcodeQuality,
+                    molecularIndexes,
+                    molecularIndexQualities);
             ret.records[1] = secondOfPair;
         }
 
         if (adapterMarker != null) {
             // Clip the read
             if (isPairedEnd) {
                 adapterMarker.adapterTrimIlluminaPairedReads(firstOfPair, secondOfPair);
-            }
-            else {
+            } else {
                 adapterMarker.adapterTrimIlluminaSingleRead(firstOfPair);
             }
         }
         return ret;
     }
+
+    private String getBarcodeQuality(ClusterData cluster) {
+        final StringJoiner barcodeQ = new StringJoiner(molecularIndexQualityDelimiter);
+
+        for (int sampleBarcodeIndex : sampleBarcodeIndices) {
+            barcodeQ.add(SAMUtils.phredToFastq(cluster.getRead(sampleBarcodeIndex).getQualities()));
+        }
+        return barcodeQ.toString();
+    }
+
+    private String getUnmatchedBarcode(ClusterData cluster, boolean withSeparator) {
+        final byte[][] barcode = new byte[sampleBarcodeIndices.length][];
+        for (int i = 0; i < sampleBarcodeIndices.length; i++) {
```

Comment:
It looks like you could make a method and call it here and in `convertClusterToOutputRecord()`, e.g.
```
byte[][] getBarcodeSeqs() {
    final byte[][] barcode = new byte[sampleBarcodeIndices.length][];
    for (int i = 0; i < sampleBarcodeIndices.length; i++) {
        barcode[i] = cluster.getRead(sampleBarcodeIndices[i]).getBases();
    }
}
```

Reply:
Oh. OK. I'll do it, though I'm not sure it comes out cleaner...